File: 53. Maximum Subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function(nums) {
    if ( nums == null ) {
        return null;
    }

    var max = -Number.MAX_VALUE;
    var sum = 0;
    var minSum = 0;

    for ( var i = 0; i < nums.length; i++ ) {
        sum += nums[i];
        max = Math.max(max, sum - minSum);
        minSum = Math.min(minSum, sum);
    }

    return max;
};

File: 53. Maximum Subarray.py
class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if nums is None or len(nums) == 0:
            return
        minS, result, sum = 0, nums[0], 0
        for num in nums:
            sum += num
            result = max(result, sum - minS)
            minS = min(minS, sum)
        return result
